File: src-tauri/src/secrets.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

const SAMUEL_DIR: &str = ".samuel";
const SECRETS_FILE: &str = "secrets.json";

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SecretsStore {
    #[serde(flatten)]
    pub entries: HashMap<String, String>,
}

pub fn secrets_path() -> Result<PathBuf, String> {
    let home = dirs::home_dir().ok_or("Cannot find home directory")?;
    let dir = home.join(SAMUEL_DIR);
    if !dir.exists() {
        fs::create_dir_all(&dir).map_err(|e| format!("Create ~/.samuel: {e}"))?;
    }
    Ok(dir.join(SECRETS_FILE))
}

pub fn load_secrets() -> SecretsStore {
    let Ok(path) = secrets_path() else {
        return SecretsStore::default();
    };
    let Ok(data) = fs::read_to_string(&path) else {
        return SecretsStore::default();
    };
    serde_json::from_str(&data).unwrap_or_default()
}

fn save_secrets(store: &SecretsStore) -> Result<(), String> {
    let path = secrets_path()?;
    let json = serde_json::to_string_pretty(store).map_err(|e| format!("JSON: {e}"))?;
    fs::write(&path, json).map_err(|e| format!("Write secrets: {e}"))
}

#[tauri::command]
pub async fn get_secret(name: String) -> Result<Option<String>, String> {
    let store = load_secrets();
    Ok(store.entries.get(&name).cloned())
}

#[tauri::command]
pub async fn set_secret(name: String, value: String) -> Result<String, String> {
    let mut store = load_secrets();
    store.entries.insert(name.clone(), value);
    save_secrets(&store)?;
    eprintln!("[secrets] stored '{name}'");
    Ok(format!("Secret '{name}' saved."))
}

#[tauri::command]
pub async fn delete_secret(name: String) -> Result<String, String> {
    let mut store = load_secrets();
    if store.entries.remove(&name).is_some() {
        save_secrets(&store)?;
        eprintln!("[secrets] deleted '{name}'");
        Ok(format!("Secret '{name}' removed."))
    } else {
        Err(format!("Secret '{name}' not found."))
    }
}

#[tauri::command]
pub async fn list_secrets() -> Result<Vec<String>, String> {
    let store = load_secrets();
    let mut names: Vec<String> = store.entries.keys().cloned().collect();
    names.sort();
    Ok(names)
}

File: electron/handlers/secrets.ts
import { readFileSync, writeFileSync, mkdirSync, existsSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";

const SAMUEL_DIR = ".samuel";
const SECRETS_FILE = "secrets.json";

type SecretsStore = Record<string, string>;

function secretsPath(): string {
	const dir = join(homedir(), SAMUEL_DIR);
	if (!existsSync(dir)) {
		mkdirSync(dir, { recursive: true });
	}
	return join(dir, SECRETS_FILE);
}

function loadSecrets(): SecretsStore {
	try {
		const path = secretsPath();
		if (!existsSync(path)) return {};
		const data = readFileSync(path, "utf-8");
		return JSON.parse(data) as SecretsStore;
	} catch {
		return {};
	}
}

function saveSecrets(store: SecretsStore): void {
	const path = secretsPath();
	writeFileSync(path, JSON.stringify(store, null, 2));
}

export async function get_secret(args: { name: string }): Promise<string | null> {
	const store = loadSecrets();
	return store[args.name] ?? null;
}

export async function set_secret(args: { name: string; value: string }): Promise<string> {
	const store = loadSecrets();
	store[args.name] = args.value;
	saveSecrets(store);
	console.error(`[secrets] stored '${args.name}'`);
	return `Secret '${args.name}' saved.`;
}

export async function delete_secret(args: { name: string }): Promise<string> {
	const store = loadSecrets();
	if (!(args.name in store)) {
		throw new Error(`Secret '${args.name}' not found.`);
	}
	delete store[args.name];
	saveSecrets(store);
	console.error(`[secrets] deleted '${args.name}'`);
	return `Secret '${args.name}' removed.`;
}

export async function list_secrets(): Promise<string[]> {
	const store = loadSecrets();
	return Object.keys(store).sort();
}
